crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -446,9 +446,19 @@ where
             continue;
         };
 
-        let last_archived_block = client
-            .block(last_archived_block_hash)?
-            .expect("Last archived block must always be retrievable; qed");
+        // If the user chooses an object mapping start block we don't have the data for, we can't
+        // create mappings for it, so the node must exit with an error.
+        let Some(last_archived_block) = client.block(last_archived_block_hash)? else {
+            let mapping = if create_object_mappings {
+                "mapping "
+            } else {
+                ""
+            };
+            let error = format!(
+                "Missing data for {mapping}block {last_archived_block_number} hash {last_archived_block_hash}"
+            );
+            return Err(sp_blockchain::Error::Application(error.into()));
+        };
```

Comment:
What about checking this in `initialize_archiver` instead? It should be sufficient to query `best_block_to_archive` there.

Reply:
This isn’t possible, because find_last_archived_block searches back through the segments until it finds a segment with a last archived block which is less than or equal to best_block_to_archive.

 initialize_archiver doesn’t know about this logic, so it can’t check the actual last_archived_block_number found by find_last_archived_block.

Reply:
Achiving has the invariant that from the tip of the chain (best block) we can traverse back to some block that allows us to initialize archiver. By overriding `best_block_to_archive` with block for which to create object mappings we constrain where to start.

Looking carefully we can see `(best_block_to_archive..best_block_number)` iteration that ensures to step up closer to the tip to find the blocks that actually exist (right now tip-100 may not actually exist in the database if we use Snap sync and with future Snap sync we may have holes in the database).

One edge case that was problematic before was genesis block, which you handled in https://github.com/autonomys/subspace/pull/3247, but that was actually not necessary strictly speaking since it is a special case of a more general problem. The same exact issue can happen if we prune blocks since Substrate still never prunes headers (Shamil is working on fixing this in https://github.com/paritytech/polkadot-sdk/pull/6451), so quick check for block existence using header in above mentioned loop will succeed, while block body is already gone.

So to handle both genesis block case after Snap sync and this, all you need to do is to add additional check after the loop that block body exists. Since the only reason for it to not exist is incorrect user input for object mapping creation CLI option, we can return a corresponding error back. In all other cases node sync maintains mentioned invariants already, hence no corresponding error handling in `find_last_archived_block` and why I'm insisting no changes need to be done there.

Moving override of `best_block_to_archive` to after the loop breaks these invariants and simply moves its handling to a different place unnecessarily.

Reply:
That makes sense, but I'm still concerned there's an edge case where the data for the chosen mapping block is present, but the data for the block at the end of the segment before the chosen block doesn't have data (which is what `find_last_archived_block` checks and can panic on).

But maybe that will never happen, and if it does we'll find it in testing.

And the user can fix that by only choosing blocks at the end of segments, or by wiping their node.

Reply:
There is actually one edge case here. We constrain pruning of state to 140k blocks because state pruning is decoupled from finalization (in contrast to pruning of block bodies), which means it is possible for block body to be present and state not be present anymore.

In practice we chose 140k blocks as something that is the most conservative case for 1 segment worth of empty blocks, in practice blocks are typically not empty (contain votes at least), so this will span more than 1 segment.

To handle this as well you may add another check for state by doing a runtime API call (might query object mapping with an empty list of extrinsics, just to see if it fails altogether), that will cover 100% of necessary checks to the best of my knowledge.